cmd/cloud-controller-manager/app/options/options.go

```diff
@@ -222,16 +210,69 @@ func (o *CloudControllerManagerOptions) Validate() error {
 	return utilerrors.NewAggregate(errors)
 }
 
+// resyncPeriod computes the time interval a shared informer waits before resyncing with the api server
+func resyncPeriod(c *cloudcontrollerconfig.Config) func() time.Duration {
+	return func() time.Duration {
+		factor := rand.Float64() + 1
+		return time.Duration(float64(c.ComponentConfig.GenericComponent.MinResyncPeriod.Nanoseconds()) * factor)
+	}
+}
+
 // Config return a cloud controller manager config objective
 func (o *CloudControllerManagerOptions) Config() (*cloudcontrollerconfig.Config, error) {
 	if err := o.Validate(); err != nil {
 		return nil, err
 	}
 
-	c := &cloudcontrollerconfig.Config{}
-	if err := o.ApplyTo(c, CloudControllerManagerUserAgent); err != nil {
+	kubeconfig, err := clientcmd.BuildConfigFromFlags(o.Master, o.Kubeconfig)
+	if err != nil {
 		return nil, err
 	}
+	kubeconfig.ContentConfig.ContentType = o.GenericComponent.ContentType
+	kubeconfig.QPS = o.GenericComponent.KubeAPIQPS
+	kubeconfig.Burst = int(o.GenericComponent.KubeAPIBurst)
+
+	client, err := clientset.NewForConfig(restclient.AddUserAgent(kubeconfig, CloudControllerManagerUserAgent))
+	if err != nil {
+		return nil, err
+	}
+
+	leaderElectionClient := clientset.NewForConfigOrDie(restclient.AddUserAgent(kubeconfig, "leader-election"))
+
+	eventRecorder := createRecorder(client, CloudControllerManagerUserAgent)
+
+	c := &cloudcontrollerconfig.Config{
+		Client:               client,
+		Kubeconfig:           kubeconfig,
+		EventRecorder:        eventRecorder,
+		LeaderElectionClient: leaderElectionClient,
+	}
+	if err := o.ApplyTo(c); err != nil {
+		return nil, err
+	}
+
+	rootClientBuilder := controller.SimpleControllerClientBuilder{
+		ClientConfig: c.Kubeconfig,
+	}
+	if c.ComponentConfig.KubeCloudShared.UseServiceAccountCredentials {
+		c.ClientBuilder = controller.SAControllerClientBuilder{
+			ClientConfig:         restclient.AnonymousClientConfig(c.Kubeconfig),
+			CoreClient:           c.Client.CoreV1(),
+			AuthenticationClient: c.Client.AuthenticationV1(),
+			Namespace:            "kube-system",
+		}
+	} else {
+		c.ClientBuilder = rootClientBuilder
+	}
+	versionedClient := rootClientBuilder.ClientOrDie("shared-informers")
+	c.InformerFactory = informers.NewSharedInformerFactory(versionedClient, resyncPeriod(c)())
+
+	// If apiserver is not running we should wait for some time and fail only then. This is particularly
+	// important when we start apiserver and controller manager at the same time.
+	err = genericcontrollermanager.WaitForAPIServer(versionedClient, 10*time.Second)
```

Comment:
this does not belong here IMO. This is plumbing code and shouldn't have anything about synchronization with other components.

Reply:
add `VersionedClient ` to restore `versionedClient` in config struct, then keep origin.